Mapping/ContainerAwareEntityListenerResolver.php

```diff
@@ -0,0 +1,115 @@
+<?php
+
+namespace Doctrine\Bundle\DoctrineBundle\Mapping;
+
+use Doctrine\ORM\Mapping\EntityListenerResolver;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+class ContainerAwareEntityListenerResolver implements EntityListenerResolver
+{
+    /**
+     * @var ContainerInterface
+     */
+    private $container;
+
+    /**
+     * @var array Map to store entity listener instances.
+     */
+    private $instances = array();
+
+    /**
+     * @var array Map to store registered service ids
+     */
+    private $serviceIds = array();
+
+    /**
+     * @param ContainerInterface $container
+     */
+    public function __construct(ContainerInterface $container)
+    {
+        $this->container = $container;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function clear($className = null)
+    {
+        if ($className === null) {
+            $this->instances = array();
+
+            return;
+        }
+
+        $className = $this->normalizeClassName($className);
+
+        if (isset($this->instances[$className])) {
+            unset($this->instances[$className]);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function register($object)
+    {
+        if ( ! is_object($object)) {
+            throw new \InvalidArgumentException(sprintf('An object was expected, but got "%s".', gettype($object)));
+        }
+
+        $className = $this->normalizeClassName(get_class($object));
+
+        $this->instances[$className] = $object;
+    }
+
+    /**
+     * @param string $className
+     * @param string $serviceId
+     */
+    public function registerService($className, $serviceId)
```

Comment:
Maybe add this to a new interface and check in compiler pass. What do you think?